In [223]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# 1.

## 1.1

In [224]:
data = pd.read_csv("./lingspam-emails.csv.bz2", sep = "\t")

In [225]:
data

spam          files                                            message
0     False    3-1msg1.txt  Subject: re : 2 . 882 s - > np np  > date : su...
1     False    3-1msg2.txt  Subject: s - > np + np  the discussion of s - ...
2     False    3-1msg3.txt  Subject: 2 . 882 s - > np np  . . . for me it ...
3     False  3-375msg1.txt  Subject: gent conference  " for the listserv "...
4     False  3-378msg1.txt  Subject: query : causatives in korean  could a...
...     ...            ...                                                ...
2888   True   spmsgc50.txt  Subject: .  international driver ' s license n...
2889   True   spmsgc51.txt  Subject: new on 95 . 8 capital fm  this is new...
2890   True   spmsgc52.txt  Subject: re : new medical technology  company ...
2891   True   spmsgc53.txt  Subject: re : your request for an overview  ye...
2892   True   spmsgc54.txt  Subject: new on capital fm  this is new at htt...

[2893 rows x 3 columns]

In [226]:
from textwrap import wrap
spam = data[data['spam'] == True]
spam_msg = spam['message'].head(5).values
for msg in spam_msg:
  print("\n".join(wrap(msg)))


Subject: great part-time or summer job !  * * * * * * * * * * * * * *
* we have display boxes with credit applications that we need to place
in the small owner-operated stores in your area . here is what you do
: 1 . introduce yourself to the store owner or manager . 2 . use our
90 % effective script which tells them how this little display box
will save their customers hundreds of dollars , be a drawing card for
their business , and make them from $ 5 . 00 to $ 15 . 00 or more for
every app sent in . 3 . find a good spot on the counter , place the
box there , and say that nothing more need be done , all you need is
his name and address so the company can send him the commission checks
. your compensaation will be $ 10 for every box you place . by
becoming a representative you could also earn a commission of $ 10 for
each application that came from that store . that is of course a much
more profitable plan , as it will pay you for months or years for a
very small effort . call 1-888 - 

In [227]:
non_spam = data[data['spam'] == False]
non_spam_msg = non_spam['message'].head(5).values
for msg in non_spam_msg:
  print("\n".join(wrap(msg)))

Subject: re : 2 . 882 s - > np np  > date : sun , 15 dec 91 02 : 25 :
02 est > from : michael < mmorse @ vm1 . yorku . ca > > subject : re :
2 . 864 queries > > wlodek zadrozny asks if there is " anything
interesting " to be said > about the construction " s > np np " . . .
second , > and very much related : might we consider the construction
to be a form > of what has been discussed on this list of late as
reduplication ? the > logical sense of " john mcnamara the name " is
tautologous and thus , at > that level , indistinguishable from " well
, well now , what have we here ? " . to say that ' john mcnamara the
name ' is tautologous is to give support to those who say that a
logic-based semantics is irrelevant to natural language . in what
sense is it tautologous ? it supplies the value of an attribute
followed by the attribute of which it is the value . if in fact the
value of the name-attribute for the relevant entity were ' chaim
shmendrik ' , ' john mcnamara the name ' would be fa

## 1.2

1.2.1
Pr_S1, Pr_S0, Pr_W1S1 == Pr(w = 1|S = 1) ,pr_w1Vs1 == Pr(w = 1 AND s = 1)

In [228]:
# 1.2.2
# y is spam
y = data.spam
y

0       False
1       False
2       False
3       False
4       False
        ...  
2888     True
2889     True
2890     True
2891     True
2892     True
Name: spam, Length: 2893, dtype: bool

In [229]:
# What is the accuracy of the naive model that predicts all emails into the majority category?
print("Majority category is:" , "Spam is", y.value_counts().idxmax(), "Frequncy:", y.value_counts().max())
print("Accuracy is", y.value_counts().max() / y.size)

Majority category is: Spam is False Frequncy: 2412
Accuracy is 0.8337366055997235


In [230]:
#1.2.3
# x is word
x = data.message.str.lower().str.contains("million").astype(int)
x

0       0
1       0
2       0
3       0
4       0
       ..
2888    0
2889    1
2890    0
2891    1
2892    1
Name: message, Length: 2893, dtype: int32

In [231]:
#1.2.4
pr_s1 = y[y == True].size / y.size
pr_s1

0.16626339440027654

In [232]:
pr_s0 = y[y == False].size / y.size
pr_s0

0.8337366055997235

## 1.3

In [233]:
# 1.3.1
pr_w1 = x.mean()
pr_w0 = 1 - pr_w1
print(pr_w1, pr_w0)


0.04839267196681645 0.9516073280331836


In [234]:
# 1.3.2
data_with_x = data.assign(contains_million = x)
data_with_x


spam          files                                            message  \
0     False    3-1msg1.txt  Subject: re : 2 . 882 s - > np np  > date : su...   
1     False    3-1msg2.txt  Subject: s - > np + np  the discussion of s - ...   
2     False    3-1msg3.txt  Subject: 2 . 882 s - > np np  . . . for me it ...   
3     False  3-375msg1.txt  Subject: gent conference  " for the listserv "...   
4     False  3-378msg1.txt  Subject: query : causatives in korean  could a...   
...     ...            ...                                                ...   
2888   True   spmsgc50.txt  Subject: .  international driver ' s license n...   
2889   True   spmsgc51.txt  Subject: new on 95 . 8 capital fm  this is new...   
2890   True   spmsgc52.txt  Subject: re : new medical technology  company ...   
2891   True   spmsgc53.txt  Subject: re : your request for an overview  ye...   
2892   True   spmsgc54.txt  Subject: new on capital fm  this is new at htt...   

      contains_million  
0                    0  
1                    0  
2                    0  
3                    0  
4                    0  
...                ...  
2888                 0  
2889                 1  
2890                 0  
2891                 1  
2892                 1  

[2893 rows x 4 columns]

In [235]:
pr_w1Vs1 = data_with_x[data_with_x['spam'] == True][data_with_x['contains_million'] == 1].size / data_with_x.size
pr_w1Vs0 = data_with_x[data_with_x['spam'] == False][data_with_x['contains_million'] == 1].size / data_with_x.size

pr_w1s1 = pr_w1Vs1 / pr_s1
pr_w0s1 = 1 - pr_w1s1
pr_w1s0 = pr_w1Vs0/ pr_s0
pr_w0s0 = 1 - pr_w1s0
print("Pr(w = 1|S = 1) is", pr_w1s1)
print("Pr(w = 1|S = 0) is", pr_w1s0)
print("Pr(w = 0|S = 1) is", pr_w0s1)
print("Pr(w = 0|S = 0) is", pr_w0s0)

Pr(w = 1|S = 1) is 0.24116424116424115
Pr(w = 1|S = 0) is 0.009950248756218905
Pr(w = 0|S = 1) is 0.7588357588357588
Pr(w = 0|S = 0) is 0.9900497512437811


C:\Users\wangs\AppData\Local\Temp\ipykernel_16524\1067333309.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  pr_w1Vs1 = data_with_x[data_with_x['spam'] == True][data_with_x['contains_million'] == 1].size / data_with_x.size
C:\Users\wangs\AppData\Local\Temp\ipykernel_16524\1067333309.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  pr_w1Vs0 = data_with_x[data_with_x['spam'] == False][data_with_x['contains_million'] == 1].size / data_with_x.size


In [236]:
# 1.3.3
pr_s1w1 = pr_w1s1 * pr_s1 / pr_w1
pr_s1w0 = pr_w0s1 * pr_s1 / pr_w0
pr_s0w1 = pr_w1s0 * pr_s0 / pr_w1
pr_s0w0 = pr_w0s0 * pr_s0 / pr_w0
print("Pr(S = 1|w = 1) is", pr_s1w1)
print("Pr(S = 1|w = 0) is", pr_s1w0)
print("Pr(S = 0|w = 1) is", pr_s0w1)
print("Pr(S = 0|w = 0) is", pr_s0w0)

Pr(S = 1|w = 1) is 0.8285714285714286
Pr(S = 1|w = 0) is 0.13258263712313842
Pr(S = 0|w = 1) is 0.17142857142857143
Pr(S = 0|w = 0) is 0.8674173628768616


1.3.4
Pr(S = 1|w = 1) and Pr(S = 0|w = 1),
Pr(S = 1|w = 0) and Pr(S = 0|w = 0) sums to one.

The reason is given the email contains word, there are only 2 possibilities which are spam or non-spam, which naturally means these two probabilities sum to 1. The same applies to the case given the email DOES NOT contains word.

## 1.4

In [237]:
predict_result = pd.Series(index=data_with_x.index, dtype='bool')

def predict(email, index):
  if (email.contains_million == 1): 
    pr_spam = pr_s1w1
  else:
    pr_spam = pr_s1w0
  predict_result.at[index] = pr_spam > 0.5


for index, email in data_with_x.iterrows():
  predict(email, index)

print(predict_result)



0       False
1       False
2       False
3       False
4       False
        ...  
2888    False
2889     True
2890    False
2891     True
2892     True
Length: 2893, dtype: bool


In [238]:
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score

cm = confusion_matrix(y, predict_result)
print("Confusion Matrix:")
print(cm)

TP = cm[1][1]
FN = cm[1][0]
FP = cm[0][1]
TN = cm[0][0]

accuracy = (TP + TN) / (TP + TN + FP + FN)
precision = TP / TP + FP ## reliability of predicted positives of all positives
recall = TP / TP + FN ## reliability of catching actual positives

# Accuracy
print(f"Accuracy: {accuracy}")

# Precision
precision = precision_score(y, predict_result)
print(f"Precision: {precision}")

# Recall
recall = recall_score(y, predict_result)
print(f"Recall: {recall}")

Confusion Matrix:
[[2388   24]
 [ 365  116]]
Accuracy: 0.8655375043207743
Precision: 0.8285714285714286
Recall: 0.24116424116424118


1.4.3
The act of calculating prior probabilites and conditional probabilities consititues model training

## 1.5


In [239]:
# 1.5.1
def predict_word(word): 
  print(f"predicting the word {word}") 
  x_word = data.message.str.lower().str.contains(word).astype(int)
  data_with_x_word = data.assign(contains_word = x_word)

  pr_w1_word = x_word.mean()
  pr_w0_word = 1 - pr_w1_word

  pr_w1Vs1_word = data_with_x_word[data_with_x_word['spam'] == True][data_with_x_word['contains_word'] == 1].size / data_with_x_word.size

  pr_w1s1_word = pr_w1Vs1_word / pr_s1
  pr_w0s1_word = 1 - pr_w1s1_word
  pr_s1w1_word = pr_w1s1_word * pr_s1 / pr_w1_word
  print("s1w1", pr_s1w1_word)
  pr_s1w0_word = pr_w0s1_word * pr_s1 / pr_w0_word
  print("s1w0",pr_s1w0_word)

  predict_result_word = pd.Series(index=data_with_x_word.index, dtype='bool')

  def predict_word(email, index):
    if (email.contains_word == 1): 
      pr_spam = pr_s1w1_word
    else:
      pr_spam = pr_s1w0_word
    predict_result_word.at[index] = pr_spam > 0.5


  for index, email in data_with_x_word.iterrows():
    predict_word(email, index)

  print("Spam Status:", predict_result_word)

  cm = confusion_matrix(y, predict_result_word)
  print("Confusion Matrix:")
  print(cm)

  TP = cm[1][1]
  FN = cm[1][0]
  FP = cm[0][1]
  TN = cm[0][0]
  print("TP", TP, "FN", FN, "FP", FP,"TN", TN)

  accuracy = (TP + TN) / (TP + TN + FP + FN)
  if (TP != 0):
    precision = TP / (TP + FP) ## reliability of predicted positives of all positives
    recall = TP / (TP + FN) ## reliability of catching actual positives

  # Accuracy
  accuracy = accuracy_score(y, predict_result_word)
  print(f"Accuracy: {accuracy}")

  if (TP != 0):
    # Precision
    precision = precision_score(y, predict_result_word)
    print(f"Precision: {precision}")

    # Recall
    recall = recall_score(y, predict_result_word)
    print(f"Recall: {recall}")



In [240]:
predict_word("conference")

predicting the word conference
s1w1 0.007396449704142012
s1w0 0.21470455570590888


C:\Users\wangs\AppData\Local\Temp\ipykernel_16524\3360603893.py:10: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  pr_w1Vs1_word = data_with_x_word[data_with_x_word['spam'] == True][data_with_x_word['contains_word'] == 1].size / data_with_x_word.size


Spam Status: 0       False
1       False
2       False
3       False
4       False
        ...  
2888    False
2889    False
2890    False
2891    False
2892    False
Length: 2893, dtype: bool
Confusion Matrix:
[[2412    0]
 [ 481    0]]
TP 0 FN 481 FP 0 TN 2412
Accuracy: 0.8337366055997235


In [241]:
predict_word("the")

predicting the word the
s1w1 0.15991471215351813
s1w0 0.39240506329114033


C:\Users\wangs\AppData\Local\Temp\ipykernel_16524\3360603893.py:10: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  pr_w1Vs1_word = data_with_x_word[data_with_x_word['spam'] == True][data_with_x_word['contains_word'] == 1].size / data_with_x_word.size


Spam Status: 0       False
1       False
2       False
3       False
4       False
        ...  
2888    False
2889    False
2890    False
2891    False
2892    False
Length: 2893, dtype: bool
Confusion Matrix:
[[2412    0]
 [ 481    0]]
TP 0 FN 481 FP 0 TN 2412
Accuracy: 0.8337366055997235


In [242]:
predict_word("hundred")

predicting the word hundred
s1w1 0.7586206896551724
s1w0 0.14151962549513866


C:\Users\wangs\AppData\Local\Temp\ipykernel_16524\3360603893.py:10: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  pr_w1Vs1_word = data_with_x_word[data_with_x_word['spam'] == True][data_with_x_word['contains_word'] == 1].size / data_with_x_word.size


Spam Status: 0       False
1       False
2       False
3        True
4       False
        ...  
2888    False
2889     True
2890    False
2891    False
2892     True
Length: 2893, dtype: bool
Confusion Matrix:
[[2384   28]
 [ 393   88]]
TP 88 FN 393 FP 28 TN 2384
Accuracy: 0.8544763221569305
Precision: 0.7586206896551724
Recall: 0.18295218295218296


1.5.2
The model's precision and accuracy is around 80% which is pretty good. However the recall is relatively low which are around 20%

1.5.3
Low recall/sensitivity means the there are a lot of false negatives, which means there are a lot of positive spam emails out there not predicted as postive. The accuracy and precision look good because the model is very selective on labeling the email as positive, resulting high precision + low recall.

1.5.4

- a. million works because it is highly associated with spam content

- b. Conference does not work because this word can relate to academic conference, which is non spam.

- c. The does not work because it is a common word among all email

- d. hundred works becuase it is highly associalted with spam content

# 2 Naive Bayes

## 2.1


In [243]:
# 2.1.1
from sklearn.feature_extraction.text import CountVectorizer
vectorizer = CountVectorizer(binary=True) # only 1/0 for the word presence!
# define vectorizer
X = vectorizer.fit_transform(data.message)
# vectorize your data. Note: this creates a sparse matrix,
# use .toarray() if you run into trouble
vocabulary = vectorizer.get_feature_names_out()
# in case you want to see what are the actual words

In [244]:
print(X)

  (0, 52792)	1
  (0, 46067)	1
  (0, 5047)	1
  (0, 40254)	1
  (0, 17080)	1
  (0, 53110)	1
  (0, 988)	1
  (0, 17248)	1
  (0, 5149)	1
  (0, 145)	1
  (0, 1970)	1
  (0, 21420)	1
  (0, 24019)	1
  (0, 37194)	1
  (0, 37769)	1
  (0, 58494)	1
  (0, 60448)	1
  (0, 12383)	1
  (0, 4992)	1
  (0, 45615)	1
  (0, 59746)	1
  (0, 60591)	1
  (0, 8775)	1
  (0, 28555)	1
  (0, 54722)	1
  :	:
  (2892, 12496)	1
  (2892, 49485)	1
  (2892, 8675)	1
  (2892, 57616)	1
  (2892, 3466)	1
  (2892, 45728)	1
  (2892, 57628)	1
  (2892, 14056)	1
  (2892, 9634)	1
  (2892, 21906)	1
  (2892, 40505)	1
  (2892, 46405)	1
  (2892, 12722)	1
  (2892, 29892)	1
  (2892, 57629)	1
  (2892, 49861)	1
  (2892, 12708)	1
  (2892, 22841)	1
  (2892, 59629)	1
  (2892, 35861)	1
  (2892, 6494)	1
  (2892, 45314)	1
  (2892, 51135)	1
  (2892, 50374)	1
  (2892, 56512)	1


In [245]:
print(f"documents number:{len(data.index)}, tokens number:{len(vocabulary)}")

documents number:2893, tokens number:60925


In [246]:
# 2.1.2a
print(f"the emails that the rows correspond to are:\n{data[982:985]}")

the emails that the rows correspond to are:
      spam           files                                            message
982  False  6-1148msg1.txt  Subject: summary : parsing of ambiguous sequen...
983  False  6-1149msg1.txt  Subject: re : sapir - whorf and what to tell s...
984  False  6-1150msg1.txt  Subject: call for contributions  call for cont...


In [247]:
# 2.1.2b
print(f"the words that the columns correspond to are:\n{vocabulary[40041:40046]}")

the words that the columns correspond to are:
['nooteboom' 'nootka' 'nope' 'nor' 'nora']


2.1.2c
The 1s in the middle of the table means:<br>
In the email with index 983, the word 'nootka' and 'nor' has 1 occurence

2.1.2d The zeros mean the corresponding words in corresponding emails does not have occurence.

In [248]:
# 2.1.3
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

2.1.4 <br>
log_pr_w1: log Pr(W = 1)<br>
l_s1w: ℓ(S = 1|W)

## 2.2

In [249]:
# 2.2.1
log_pr_s1 = np.log(y_train[y_train == 1].size / y_train.size)
log_pr_s0 = np.log(y_train[y_train == 0].size / y_train.size)
print(log_pr_s0, log_pr_s1)

-0.18717341174592203 -1.7678471428525


In [250]:
# 2.2.2
N_s1 = len(y_train[y_train == 1])
N_s0 = len(y_train[y_train == 0])

N_w1s1 = np.sum(X_train[y_train == 1], axis=0)
N_w1s0 = np.sum(X_train[y_train == 0], axis=0)

log_pr_w1s1 = np.array(np.log(N_w1s1/ N_s1)).flatten()
log_pr_w1s0 = np.array(np.log(N_w1s0 / N_s0)).flatten()

print(log_pr_w1s1)
print(log_pr_w1s0)


[-1.11907336 -1.14260386 -5.97888576 ...        -inf -5.97888576
        -inf]
[-1.77573431 -3.0056826  -7.5595595  ... -7.5595595         -inf
        -inf]


C:\Users\wangs\AppData\Local\Temp\ipykernel_16524\3517566930.py:8: RuntimeWarning: divide by zero encountered in log
  log_pr_w1s1 = np.array(np.log(N_w1s1/ N_s1)).flatten()
C:\Users\wangs\AppData\Local\Temp\ipykernel_16524\3517566930.py:9: RuntimeWarning: divide by zero encountered in log
  log_pr_w1s0 = np.array(np.log(N_w1s0 / N_s0)).flatten()


The dimensions should correspond to the number of unique words because they contain the probabilities of each unique word's occurence

In [251]:
# 2.3.1
l_s1w1 = log_pr_s1 + X_test.dot(log_pr_w1s1)
l_s0w1 = log_pr_s0 + X_test.dot(log_pr_w1s0)
print(l_s1w1.shape, l_s0w1.shape)


(579,) (579,)


In [252]:
X_test.shape

(579, 60925)

2.3.2
I computed 579 log-likelihood because each email has different occurence of words therefore the sum of logPr(w|S=1) is different in the log-likelihood equation. Therefore I need to compute different log-likelihood for each email

In [253]:
predict_result_naive = pd.Series(index=np.arange(X_test.shape[0]), dtype='bool')

for index in [0, X_test.shape[0] - 1]:
  if l_s1w1[index] > l_s0w1[index]:
    predict_result_naive[index] = True
  else: predict_result_naive[index] = False

print(predict_result_naive)
predict_result_naive.value_counts()

0      False
1       True
2       True
3       True
4       True
       ...  
574     True
575     True
576     True
577     True
578    False
Length: 579, dtype: bool


True     577
False      2
Name: count, dtype: int64

In [254]:
cm = confusion_matrix(y_test, predict_result_naive)
print("Confusion Matrix:")
print(cm)

TP = cm[1][1]
FN = cm[1][0]
FP = cm[0][1]
TN = cm[0][0]
print("TP", TP, "FN", FN, "FP", FP,"TN", TN)

accuracy = (TP + TN) / (TP + TN + FP + FN)
if (TP != 0):
  precision = TP / (TP + FP) ## reliability of predicted positives of all positives
  recall = TP / (TP + FN) ## reliability of catching actual positives

# Accuracy
accuracy = accuracy_score(y_test, predict_result_naive)
print(f"Accuracy: {accuracy}")

if (TP != 0):
  # Precision
  precision = precision_score(y_test, predict_result_naive)
  print(f"Precision: {precision}")

  # Recall
  recall = recall_score(y_test, predict_result_naive)
  print(f"Recall: {recall}")

Confusion Matrix:
[[  2 491]
 [  0  86]]
TP 86 FN 0 FP 491 TN 2
Accuracy: 0.15198618307426598
Precision: 0.14904679376083188
Recall: 1.0


2.3.5 With no smoothing, the model will label the email as spam even if there is only one occurence of a word corresponding to spam